File: typescript/problem_6.ts
// XorLinkedList is an XOR doubly-linked list backed by an array.
// Array indices serve as stand-in memory addresses.
interface XorNodeEntry { val: number; both: number; }

export class XorLinkedList {
    private nodes: XorNodeEntry[] = [];
    private head = -1;
    private tail = -1;

    // add appends an element to the end of the list.
    add(val: number): void {
        const idx = this.nodes.length;
        const prev = this.tail;
        this.nodes.push({ val, both: prev ^ -1 }); // both = prev XOR null(-1)
        if (this.tail !== -1) {
            // update old tail: replace null(-1) with new index
            this.nodes[this.tail].both ^= (-1 ^ idx);
        }
        this.tail = idx;
        if (this.head === -1) this.head = idx;
    }

    // get returns the value at the given 0-based index.
    get(index: number): number {
        let prev = -1, curr = this.head;
        for (let i = 0; i < index; i++) {
            const next = prev ^ this.nodes[curr].both; // XOR out prev to get next
            prev = curr;
            curr = next;
        }
        return this.nodes[curr].val;
    }
}

export function problem6(): void {
    console.log("Problem 6:");
    const lst = new XorLinkedList();
    for (const v of [1, 2, 3, 4]) lst.add(v);
    if (lst.get(0) !== 1) throw new Error("assertion failed");
    if (lst.get(1) !== 2) throw new Error("assertion failed");
    if (lst.get(2) !== 3) throw new Error("assertion failed");
    if (lst.get(3) !== 4) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_6.rs
// XorLinkedList is an XOR doubly-linked list backed by a Vec.
// Vec indices serve as stand-in memory addresses; usize::MAX is null.

const NULL: usize = usize::MAX;

struct XorNode { val: i64, both: usize }

pub struct XorLinkedList {
    nodes: Vec<XorNode>,
    head: usize,
    tail: usize,
}

impl XorLinkedList {
    pub fn new() -> Self {
        Self { nodes: Vec::new(), head: NULL, tail: NULL }
    }

    // add appends an element to the end of the list.
    pub fn add(&mut self, val: i64) {
        let idx = self.nodes.len();
        let prev = self.tail;
        // both = prev XOR null; XOR in Rust: prev ^ NULL
        self.nodes.push(XorNode { val, both: prev ^ NULL });
        if self.tail != NULL {
            // update old tail: replace NULL with new index
            self.nodes[self.tail].both ^= NULL ^ idx;
        }
        self.tail = idx;
        if self.head == NULL { self.head = idx; }
    }

    // get returns the value at the given 0-based index.
    pub fn get(&self, index: usize) -> i64 {
        let mut prev = NULL;
        let mut curr = self.head;
        for _ in 0..index {
            let next = prev ^ self.nodes[curr].both; // XOR out prev to get next
            prev = curr;
            curr = next;
        }
        self.nodes[curr].val
    }
}

pub fn problem6() {
    println!("Problem 6:");
    let mut lst = XorLinkedList::new();
    for v in [1, 2, 3, 4] { lst.add(v); }
    assert_eq!(lst.get(0), 1, "assertion failed");
    assert_eq!(lst.get(1), 2, "assertion failed");
    assert_eq!(lst.get(2), 3, "assertion failed");
    assert_eq!(lst.get(3), 4, "assertion failed");
    println!("All Tests Passed");
}
